app/assets/javascripts/test/puppeteer/dataset_rendering_helpers.js

```diff
@@ -1,46 +1,62 @@
+// @flow
 /* eslint import/no-extraneous-dependencies: ["error", {"peerDependencies": true}], no-await-in-loop: 0 */
-const urljoin = require("url-join");
-const fetch = require("node-fetch");
-const pixelmatch = require("pixelmatch");
+import urljoin from "url-join";
+import pixelmatch from "pixelmatch";
+import mergeImg from "merge-img";
+import type { Page } from "puppeteer";
+import { createExplorational } from "../../admin/admin_rest_api";
 
-const DEV_AUTH_TOKEN = "secretScmBoyToken";
+export const DEV_AUTH_TOKEN = "secretScmBoyToken";
 
-async function createExplorational(datasetName, typ, withFallback, baseUrl) {
-  const fullUrl = urljoin(baseUrl, `/api/datasets/${datasetName}/createExplorational`);
-  return (await fetch(fullUrl, {
-    body: JSON.stringify({ typ, withFallback }),
-    method: "POST",
+type Screenshot = {
+  screenshot: Buffer,
+  width: number,
+  height: number,
+};
+
+function getDefaultRequestOptions(baseUrl: string) {
+  return {
+    host: baseUrl,
     headers: {
       "X-Auth-Token": DEV_AUTH_TOKEN,
-      "Content-Type": "application/json",
     },
-  })).json();
+  };
 }
 
-async function screenshotDataset(page, baseUrl, datasetName) {
-  const createdExplorational = await createExplorational(datasetName, "skeleton", false, baseUrl);
+export async function screenshotDataset(
+  page: Page,
+  baseUrl: string,
+  datasetName: string,
+): Promise<Screenshot> {
+  const options = getDefaultRequestOptions(baseUrl);
+  const createdExplorational = await createExplorational(datasetName, "skeleton", false, options);
   return openTracingViewAndScreenshot(page, baseUrl, createdExplorational.id);
 }
 
-async function openTracingViewAndScreenshot(page, baseUrl, annotationId) {
-  await page.goto(urljoin(baseUrl, `/annotations/Explorational/${annotationId}`), {
-    timeout: 0,
-  });
-
-  let canvas;
-  while (canvas == null) {
-    canvas = await page.$("#render-canvas");
+async function waitForTracingViewLoad(page: Page) {
+  let inputCatchers;
+  while (inputCatchers == null || inputCatchers.length < 4) {
+    inputCatchers = await page.$(".inputcatcher");
     await page.waitFor(500);
   }
-  const { width, height } = await canvas.boundingBox();
+}
+
+async function waitForRenderingFinish(page: Page): Promise<Buffer> {
+  const tdViewSelector = "#inputcatcher_TDView";
```

Comment:
Is this reliable with respect to bucket loading? Since the td view never displays all bucket data simultaneously (there is always some occlusion), I can imagine cases in which the td view will not change, but the others will.

Can't we still screenshot the canvas to see whether the page is ready and then screenshot the individual parts for diffing?

Reply:
I wasn't sure about screenshotting the canvas, because the FPS meter keeps changing for a while even after all data is rendered. I agree with you though, and will try to screenshot the page without the FPS meter or to remove the FPS meter from the DOM before :)